<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · VCFTools</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">VCFTools</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../summaryinfo/">VCF Summary</a></li><li><a class="tocitem" href="../filter/">Filter</a></li><li><a class="tocitem" href="../convert/">Convert</a></li><li><a class="tocitem" href="../write/">Write</a></li><li><a class="tocitem" href="../conformgt/">Match markers in two VCF</a></li><li><a class="tocitem" href="../grm/">Genetic Relationship Matrix</a></li><li><a class="tocitem" href="../aim/">Ancestry informative markers</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/VCFTools.jl/blob/master/docs/src/man/api.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h1><p>Documentation for <code>VCFTools.jl</code>&#39;s functions.</p><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#VCFTools.aim_select"><code>VCFTools.aim_select</code></a></li><li><a href="#VCFTools.conformgt_by_id"><code>VCFTools.conformgt_by_id</code></a></li><li><a href="#VCFTools.conformgt_by_pos"><code>VCFTools.conformgt_by_pos</code></a></li><li><a href="#VCFTools.convert_ds"><code>VCFTools.convert_ds</code></a></li><li><a href="#VCFTools.convert_gt"><code>VCFTools.convert_gt</code></a></li><li><a href="#VCFTools.convert_ht"><code>VCFTools.convert_ht</code></a></li><li><a href="#VCFTools.copy_ds!"><code>VCFTools.copy_ds!</code></a></li><li><a href="#VCFTools.copy_gt!"><code>VCFTools.copy_gt!</code></a></li><li><a href="#VCFTools.copy_ht!"><code>VCFTools.copy_ht!</code></a></li><li><a href="#VCFTools.filter_genotype"><code>VCFTools.filter_genotype</code></a></li><li><a href="#VCFTools.filter_header"><code>VCFTools.filter_header</code></a></li><li><a href="#VCFTools.gtstats"><code>VCFTools.gtstats</code></a></li><li><a href="#VCFTools.mask_gt"><code>VCFTools.mask_gt</code></a></li><li><a href="#VCFTools.nrecords"><code>VCFTools.nrecords</code></a></li><li><a href="#VCFTools.nsamples"><code>VCFTools.nsamples</code></a></li><li><a href="#VCFTools.openvcf"><code>VCFTools.openvcf</code></a></li><li><a href="#VCFTools.sampleID"><code>VCFTools.sampleID</code></a></li><li><a href="#VCFTools.write_vcf"><code>VCFTools.write_vcf</code></a></li></ul><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="VCFTools.openvcf" href="#VCFTools.openvcf"><code>VCFTools.openvcf</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">openvcf(vcffile, [mode = &quot;r&quot;])</code></pre><p>Open VCF file (<code>.vcf</code> or <code>.vcf.gz</code>) and return an IO stream.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/gtstats.jl#L250-L254">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.nrecords" href="#VCFTools.nrecords"><code>VCFTools.nrecords</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nrecords(vcffile)</code></pre><p>Number of records (markers) in a VCF file. Each record is a row. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/gtstats.jl#L269-L273">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.nsamples" href="#VCFTools.nsamples"><code>VCFTools.nsamples</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nsamples(vcffile)</code></pre><p>Number of samples (individuals) in a VCF file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/gtstats.jl#L285-L289">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.sampleID" href="#VCFTools.sampleID"><code>VCFTools.sampleID</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sampleID(vcffile::AbstractString)</code></pre><p>Helper function for extracting sample IDs from a VCF file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/gtstats.jl#L297-L301">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.gtstats" href="#VCFTools.gtstats"><code>VCFTools.gtstats</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">gtstats(vcffile, [out=devnull])</code></pre><p>Calculate genotype statistics for each marker with GT field in a VCF file.</p><p><strong>Input</strong></p><ul><li><code>vcffile</code>: VCF file, ending with <code>.vcf</code> or <code>.vcf.gz</code></li><li><code>out</code>: output file name or IOStream. Default is <code>out=devnull</code> (no output).</li></ul><p>One line with 15 tab-delimited fiels is written per marker to <code>out</code>:     - 1-8)  VCF fixed fields (CHROM, POS, ID, REF, ALT, QUAL, FILT, INFO)     -   9)  Missing genotype count     -  10)  Missing genotype frequency     -  11)  ALT allele count     -  12)  ALT allele frequency     -  13)  Minor allele count             (REF allele vs ALT alleles)     -  14)  Minor allele frequency         (REF allele vs ALT alleles)     -  15)  HWE P-value                    (REF allele vs ALT alleles)</p><ul><li><code>pval_method</code>: Method for calculating hardy weinburg equilibrium p-values.    Can be <code>:Pearson</code> (default) or <code>:Fisher</code>.</li></ul><p><strong>Output</strong></p><ul><li><code>records</code>: number of records in the input VCF file</li><li><code>samples</code>: number of individuals in the input VCF file</li><li><code>lines</code>  : number of lines written to <code>out</code>; equivalently number of markers   with GT field</li><li><code>missings_by_sample</code>: number of missing genotypes in each sample</li><li><code>missings_by_record</code>: number of missing genotypes in each record (marker)</li><li><code>maf_by_record</code>: minor allele frequency in each record (marker)</li><li><code>minorallele_by_record</code>: a Boolean vector indicating the minor allele in each   record (marker). <code>minorallele_by_record[i]=true</code> means the minor allele is   the REF allele for marker <code>i</code>; <code>minorallele_by_record[i]=false</code> means the   minor allele is the ALT allele for marker <code>i</code></li><li><code>hwe_by_record</code>: hardy weinburg equilibrium p-values in each record calculated   using <code>pval_method</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/gtstats.jl#L1-L35">source</a></section><section><div><pre><code class="nohighlight hljs">gtstats(record, [missings_by_sample=nothing])</code></pre><p>Calculate genotype statistics for a VCF record with GT field.</p><p><strong>Input</strong></p><ul><li><code>record</code>: a VCF record</li><li><code>missings_by_sample</code>: accumulator of missings by sample, <code>missings_by_sample[i]</code> is incremented by 1 if <code>i</code>-th individual has missing genotype in this record</li></ul><p><strong>Output</strong></p><ul><li><code>n00</code>: number of homozygote REF/REF or REF|REF</li><li><code>n01</code>: number of heterozygote REF/ALT or REF|ALT</li><li><code>n11</code>: number of homozygote ALT/ALT or ALT|ALT</li><li><code>n0</code>: number of REF alleles</li><li><code>n1</code>: number of ALT alleles</li><li><code>altfreq</code>: proportion of ALT alleles</li><li><code>reffreq</code>: proportion of REF alleles</li><li><code>missings</code>: number of missing genotypes</li><li><code>minorallele</code>: minor allele: <code>false</code> (ALT is minor allele) or <code>true</code> (REF is minor allele)</li><li><code>maf</code>: minor allele frequency</li><li><code>hwepval</code>: Hardy-Weinberg p-value</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/gtstats.jl#L92-L113">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>filter</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>filter_record!</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="VCFTools.filter_header" href="#VCFTools.filter_header"><code>VCFTools.filter_header</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_header(reader, sample_mask)</code></pre><p>Filters the header info of a VCF reader. Will not keep sample IDs for entry <code>i</code> if <code>sample_mask[i] = false</code>.</p><p><strong>Inpute:</strong></p><ul><li><code>reader</code>: a VCF reader object.</li><li><code>sample_mask</code>: a BitVector. <code>sample_mask[i] = true</code> means keep sample <code>i</code>. </li></ul><p>TODO: Create VCFTool.jl signature in meta info. If filedate/signature exist already, delete them before writing new ones</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/filter.jl#L81-L91">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.filter_genotype" href="#VCFTools.filter_genotype"><code>VCFTools.filter_genotype</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_genotype(record, [genokey=[&quot;GT&quot;]])</code></pre><p>Filter a VCF record according to <code>genokey</code> and output a VCF record with genotype formats only in <code>genokey</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/conformgt.jl#L266-L271">source</a></section><section><div><pre><code class="nohighlight hljs">filter_genotype(vcffile, outfile, [genokey=[&quot;GT&quot;]])</code></pre><p>Filter a VCF file according to <code>genokey</code> (default <code>[&quot;GT&quot;]</code>). Output a VCF file with genotype formats only in <code>genokey</code>. Record (markers) with no fields in <code>genokey</code> are skipped.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/conformgt.jl#L283-L289">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.convert_gt" href="#VCFTools.convert_gt"><code>VCFTools.convert_gt</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Convert a two-bit genotype (of ALT alleles) to a real number  of type <code>t</code> according to specified SNP model. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/convert.jl#L1-L5">source</a></section><section><div><pre><code class="nohighlight hljs">convert_gt(t, vcffile; [impute=false], [center=false], [scale=false])</code></pre><p>Convert the GT data from a VCF file to a matrix of type <code>t</code>, allowing for  missing values. Records with missing genotypes are converted to <code>missing</code>.</p><p><strong>Input</strong></p><ul><li><code>t</code>: a type <code>t &lt;: Real</code></li><li><code>vcffile</code>: VCF file path</li></ul><p><strong>Optional argument</strong></p><ul><li><code>model</code>: genetic model <code>:additive</code> (default), <code>:dominant</code>, or <code>:recessive</code></li><li><code>impute</code>: impute missing genotype or not, default <code>false</code></li><li><code>center</code>: center gentoype by 2maf or not, default <code>false</code></li><li><code>scale</code>: scale genotype by 1/√2maf(1-maf) or not, default <code>false</code></li><li><code>trans</code>: whether to import data transposed so that each column is 1 genotype,   default <code>false</code></li><li><code>msg</code>: A message that will be printed to indicate progress. Defaults to   not printing. </li><li><code>save_snp_info</code>: Boolean. If true, will also output sample ID, chrom, pos,   snp id, ref, and alt vectors</li></ul><p><strong>Output</strong></p><ul><li><code>out</code>: The genotype matrix stored in column or rows depending on <code>trans</code>.</li><li><code>sampleID</code>: The sample ID for each person</li><li><code>record_chr</code>: The chromosome number for each record</li><li><code>record_pos</code>: The position for each record</li><li><code>record_ids</code>: The record ID for each record. Only the first one is   recorded when there are multiple IDs. </li><li><code>record_ref</code>: The ref allele for each record</li><li><code>record_alt</code>: The alternate allele for each record. Only the first one is   recorded when there are multiple alternate alleles. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/convert.jl#L229-L261">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.convert_ht" href="#VCFTools.convert_ht"><code>VCFTools.convert_ht</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">convert_ht(t, vcffile)</code></pre><p>Converts the GT data from a VCF file to a haplotype matrix of type <code>t</code>. One  VCF record will become 2 column/row in the resulting matrix. Records cannot have missing genotypes and the target file must be phased. </p><p><strong>Input</strong></p><ul><li><code>t</code>: a type <code>t &lt;: Real</code>. If <code>t</code> is <code>Bool</code>, output will be a BitMatrix.</li><li><code>vcffile</code>: a phased VCF file</li><li><code>trans</code>: whether to import data transposed so that each column is a   haplotype, default <code>false</code>.</li><li><code>msg</code>: A message that will be printed to indicate import progress. Defaults to   not printing. </li><li><code>save_snp_info</code>: Boolean. If true, will also output sample ID, chrom, pos,   snp id, ref, and alt vectors</li></ul><p><strong>Output</strong></p><ul><li><code>out</code>: The haplotype matrix stored in column or rows depending on <code>trans</code>.</li><li><code>sampleID</code>: The sample ID for each person</li><li><code>record_chr</code>: The chromosome number for each record</li><li><code>record_pos</code>: The position for each record</li><li><code>record_ids</code>: The record ID for each record. Only the first one is   recorded when there are multiple IDs. </li><li><code>record_ref</code>: The ref allele for each record</li><li><code>record_alt</code>: The alternate allele for each record. Only the first one is   recorded when there are multiple alternate alleles. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/convert.jl#L305-L332">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.convert_ds" href="#VCFTools.convert_ds"><code>VCFTools.convert_ds</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">convert_ds(t, vcffile; [key = &quot;DS&quot;], ...)</code></pre><p>Converts dosage data from a VCF file to a numeric matrix of type <code>t</code>. Here <code>key</code> specifies the FORMAT field of the VCF file that encodes the dosage (default = &quot;DS&quot;). </p><p><strong>Arguments</strong></p><ul><li><code>t</code>: type of output matrix. </li><li><code>vcffile</code>: VCF file path</li></ul><p><strong>Optional argument</strong></p><ul><li><code>key</code>: The FIELD name of the VCF file that encodes dosages, defaults to <code>&quot;DS&quot;</code></li><li><code>model</code>: genetic model <code>:additive</code> (default), <code>:dominant</code>, or <code>:recessive</code></li><li><code>impute</code>: impute missing genotype with 2 times the ALT allele frequency or   not. Default <code>false</code></li><li><code>center</code>: center gentoype by 2maf or not, default <code>false</code></li><li><code>scale</code>: scale genotype by 1/√2maf(1-maf) or not, default <code>false</code></li><li><code>trans</code>: whether to import data transposed so that each column is 1 genotype,   default <code>false</code></li><li><code>msg</code>: A message that will be printed to indicate progress. Defaults to   not printing. </li></ul><p><strong>Output</strong></p><ul><li><code>out</code>: Dosage matrix </li><li><code>sampleID</code>: The sample ID for each person</li><li><code>record_chr</code>: The chromosome number for each record</li><li><code>record_pos</code>: The position for each record</li><li><code>record_ids</code>: The record ID for each record. Only the first one is   recorded when there are multiple IDs. </li><li><code>record_ref</code>: The ref allele for each record</li><li><code>record_alt</code>: The alternate allele for each record. Only the first one is   recorded when there are multiple alternate alleles. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/convert.jl#L545-L578">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.copy_gt!" href="#VCFTools.copy_gt!"><code>VCFTools.copy_gt!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">copy_gt!(A, reader; ...)</code></pre><p>Fill the rows of matrix <code>A</code> by the GT data from VCF records in <code>reader</code>. Records with missing genotypes are converted to <code>missing</code>.</p><p><strong>Input</strong></p><ul><li><code>A</code>: a matrix or vector such that <code>eltype(A) &lt;: Union{Missing, Real}</code>. Each   row is a person&#39;s genotype. </li><li><code>reader</code>: a VCF reader</li></ul><p><strong>Optional argument</strong></p><ul><li><code>model</code>: genetic model <code>:additive</code> (default), <code>:dominant</code>, or <code>:recessive</code></li><li><code>impute</code>: impute missing genotype or not, default <code>false</code></li><li><code>center</code>: center gentoype by 2maf or not, default <code>false</code></li><li><code>scale</code>: scale genotype by 1/√2maf(1-maf) or not, default <code>false</code></li><li><code>msg</code>: A message that will be printed to indicate progress. Defaults to   not printing. </li><li><code>sampleID</code>: The sample ID for each person</li><li><code>record_chr</code>: The chromosome number for each record</li><li><code>record_pos</code>: The position for each record</li><li><code>record_ids</code>: The record ID for each record. Only the first one is   recorded when there are multiple IDs. </li><li><code>record_ref</code>: The ref allele for each record</li><li><code>record_alt</code>: The alternate allele for each record. Only the first one is   recorded when there are multiple alternate alleles. </li></ul><p><strong>Output</strong></p><ul><li><code>A</code>: <code>ismissing(A[i, j]) == true</code> indicates missing genotype. If <code>impute=true</code>,   <code>ismissing(A[i, j]) == false</code> for all entries.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/convert.jl#L21-L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.copy_ht!" href="#VCFTools.copy_ht!"><code>VCFTools.copy_ht!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">copy_ht!(A, reader)</code></pre><p>Fill 2 columns of <code>A</code> by the GT data from VCF records in <code>reader</code>, each record filling 2 columns. Missing data is not allowed and genotypes must be phased. </p><p><strong>Input</strong></p><ul><li><code>A</code>: matrix where rows are haplotypes. Person <code>i</code>&#39;s haplotype are filled in   rows <code>2i - 1</code> and <code>2i</code>. </li><li><code>reader</code>: a VCF reader</li></ul><p><strong>Optional inputs</strong></p><ul><li><code>msg</code>: A message that will be printed to indicate progress. Defaults to   not printing.</li><li><code>sampleID</code>: The sample ID for each person</li><li><code>record_chr</code>: The chromosome number for each record</li><li><code>record_pos</code>: The position for each record</li><li><code>record_ids</code>: The record ID for each record. Only the first one is   recorded when there are multiple IDs. </li><li><code>record_ref</code>: The ref allele for each record</li><li><code>record_alt</code>: The alternate allele for each record. Only the first one is   recorded when there are multiple alternate alleles. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/convert.jl#L371-L393">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.copy_ds!" href="#VCFTools.copy_ds!"><code>VCFTools.copy_ds!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">copy_ds!(A, reader; [key=&quot;DS&quot;], ...)</code></pre><p>Fill the columns of matrix <code>A</code> by the dosage data from VCF records in <code>reader</code>. Records with missing fields (i.e. <code>key</code> field is <code>.</code>) are converted to <code>missing</code>.</p><p><strong>Input</strong></p><ul><li><code>A</code>: matrix where rows are dosages. </li><li><code>reader</code>: a VCF reader</li></ul><p><strong>Optional argument</strong></p><ul><li><code>key</code>: The field key for accessing dosage values, default <code>&quot;DS&quot;</code></li><li><code>model</code>: genetic model <code>:additive</code> (default), <code>:dominant</code>, or <code>:recessive</code></li><li><code>impute</code>: impute missing genotype or not, default <code>false</code></li><li><code>center</code>: center gentoype by 2maf or not, default <code>false</code></li><li><code>scale</code>: scale genotype by 1/√2maf(1-maf) or not, default <code>false</code></li><li><code>msg</code>: A message that will be printed to indicate progress. Defaults to   not printing.</li><li><code>sampleID</code>: The sample ID for each person</li><li><code>record_chr</code>: The chromosome number for each record</li><li><code>record_pos</code>: The position for each record</li><li><code>record_ids</code>: The record ID for each record. Only the first one is   recorded when there are multiple IDs. </li><li><code>record_ref</code>: The ref allele for each record</li><li><code>record_alt</code>: The alternate allele for each record. Only the first one is   recorded when there are multiple alternate alleles. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/convert.jl#L623-L650">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.conformgt_by_id" href="#VCFTools.conformgt_by_id"><code>VCFTools.conformgt_by_id</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">conformgt_by_id(reffile, tgtfile, outfile, chrom, posrange, checkfreq)</code></pre><p>Match the VCF records in <code>tgtfile</code> to those in <code>reffile</code> according to ID. The function will:</p><ol><li>Find corresponding VCF records in the target and reference files</li><li>Exclude target VCF records whose ID cannot be matched to any reference VCF record</li><li>Exclude target VCF records whose test of equal allele frequency is rejected at significance level <code>checkfreq</code></li><li>Adjust target VCF records so that chromosome strand and allele order match the VCF reference file</li><li>The matched VCF records are written into files <code>outfile.tgt.vcf.gz</code> and</li></ol><p><code>outfile.ref.vcf.gz</code>, both with only &quot;GT&quot; data</p><p><strong>Input</strong></p><ul><li><code>reffile</code>: VCF file with reference genotype (GT) data</li><li><code>tgtfile</code>: VCF file with target genotype (GT) data</li><li><code>outfile</code>: the prefix for output filenames</li><li><code>chrom</code>: chromosome name, must be identical in target and reference files</li><li><code>posrange</code>: position range in the reference file</li><li><code>checkfreq</code>: significance level for testing equal alelle frequencies between mached target and reference records. If the test pvalue is  <code>≤ checkfreq</code>, the records are not output. Setting <code>checkfreq=0</code> or <code>checkfreq=false</code> (default) implies not checking allele frequencies. Setting <code>checkfreq=1</code> effectively rejects all tests and no matched records are output</li></ul><p><strong>Output</strong></p><ul><li><code>lines</code>: number of matched VCF records</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/conformgt.jl#L1-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.conformgt_by_pos" href="#VCFTools.conformgt_by_pos"><code>VCFTools.conformgt_by_pos</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">conformgt_by_pos(reffile, tgtfile, outfile, chrom, posrange, checkfreq)</code></pre><p>Match the VCF GT records in <code>tgtfile</code> to those in <code>reffile</code> according to chromosome position. The function will:</p><ol><li>Find corresponding VCF records in the target and reference files</li><li>Exclude target VCF records whose position cannot be matched to any reference VCF record</li><li>Exclude target VCF records whose test of equal allele frequency is rejected at significance level <code>checkfreq</code></li><li>Adjust target VCF records so that chromosome strand and allele order match the VCF reference file</li><li>The matched VCF records are written into files <code>outfile.tgt.vcf.gz</code> and</li></ol><p><code>outfile.ref.vcf.gz</code>, both with only &quot;GT&quot; data</p><p><strong>Input</strong></p><ul><li><code>reffile</code>: VCF file with reference genotype (GT) data</li><li><code>tgtfile</code>: VCF file with target genotype (GT) data</li><li><code>outfile</code>: the prefix for output filenames</li><li><code>chrom</code>: chromosome name, must be identical in target and reference files</li><li><code>posrange</code>: position range in the reference file</li><li><code>checkfreq</code>: significance level for testing equal alelle frequencies between mached target and reference records. If the test pvalue is  <code>≤ checkfreq</code>, the records are not output. Setting <code>checkfreq=0</code> or <code>checkfreq=false</code> (default) implies not checking allele frequencies. Setting <code>checkfreq=1</code> effectively rejects all tests and no matched records are output</li></ul><p><strong>Output</strong></p><ul><li><code>lines</code>: number of matched VCF records</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/conformgt.jl#L128-L150">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.mask_gt" href="#VCFTools.mask_gt"><code>VCFTools.mask_gt</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">mask_gt(src, masks; [des = &quot;masked.&quot; * src], [separator = &#39;/&#39;])</code></pre><p>Creates a new VCF file <code>des</code> where genotype entry (i, j) of <code>src</code>  is missing if <code>masks[i, j]</code> is true. <code>src</code> is unchanged.</p><p><strong>Arguments</strong></p><ul><li><code>src</code>: Input VCF file name.</li><li><code>masks</code>: Bit matrix. <code>masks[i, j] = true</code> means mask entry (i, j).</li><li><code>des</code>: output VCF file name.</li><li><code>unphase</code>: If <code>true</code>, all heterozygous genotypes will be <code>1/0</code> and homozygous separator will be <code>/</code>. If <code>false</code>, all unmasked genotypes will be retained.  </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/filter.jl#L168-L179">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.write_vcf" href="#VCFTools.write_vcf"><code>VCFTools.write_vcf</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">write_vcf(outfile, x1, x2, [chr], [pos], [ref], [alt], [sampleID], [SNPids])</code></pre><p>Writes haplotypes in <code>x1</code> and <code>x2</code> into VCF file, tracking phase information. </p><p><strong>Inputs</strong></p><ul><li><code>outfile</code>: Output file name (ends in <code>.vcf</code> or <code>.vcf.gz</code>)</li><li><code>x1</code>: <code>n × p</code> matrix of the 1st haplotype for each sample. Each row is a haplotype,    and each entry is 0 or 1</li><li><code>x2</code>: <code>n × p</code> matrix of the 2nd haplotype for each sample. <code>x = x1 + x2</code>.    Each entry is 0 or 1</li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>chr</code>: Vector of String. <code>chr[i]</code> is chromosome number of SNP <code>i</code>. </li><li><code>pos</code>: Vector of Int. <code>pos[i]</code> is position of SNP <code>i</code>. </li><li><code>ref</code>: Vector of String. <code>ref[i]</code> is reference allele of SNP <code>i</code>. </li><li><code>alt</code>: Vector of String. <code>alt[i]</code> is alternate allele of SNP <code>i</code>. </li><li><code>sampleID</code>: Vector of String. <code>sampleID[i]</code> is reference allele of SNP <code>i</code>. </li><li><code>SNPids</code>: Vector of String. <code>SNPids[i]</code> is reference allele of SNP <code>i</code>. </li></ul><p><strong>Reference</strong></p><p>For multithreaded write, see https://github.com/OpenMendel/MendelImpute.jl/blob/master/src/impute.jl#L23</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/write.jl#L1-L24">source</a></section><section><div><pre><code class="nohighlight hljs">write_vcf(outfile, x, [chr], [pos], [ref], [alt], [sampleID], [SNPids])</code></pre><p>Writes genotypes in <code>x</code> into VCF file, not tracking phase information. </p><p><strong>Inputs</strong></p><ul><li><code>outfile</code>: Output file name (ends in <code>.vcf</code> or <code>.vcf.gz</code>)</li><li><code>x</code>: <code>n × p</code> genotype matrix. Each row is a sample, and each entry is 0, 1, or 2. </li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>chr</code>: Vector of String. <code>chr[i]</code> is chromosome number of SNP <code>i</code>. </li><li><code>pos</code>: Vector of Int. <code>pos[i]</code> is position of SNP <code>i</code>. </li><li><code>ref</code>: Vector of String. <code>ref[i]</code> is reference allele of SNP <code>i</code>. </li><li><code>alt</code>: Vector of String. <code>alt[i]</code> is alternate allele of SNP <code>i</code>. </li><li><code>sampleID</code>: Vector of String. <code>sampleID[i]</code> is reference allele of SNP <code>i</code>. </li><li><code>SNPids</code>: Vector of String. <code>SNPids[i]</code> is reference allele of SNP <code>i</code>. </li></ul><p><strong>Reference</strong></p><p>For multithreaded write, see https://github.com/OpenMendel/MendelImpute.jl/blob/master/src/impute.jl#L23</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/write.jl#L96-L116">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="VCFTools.aim_select" href="#VCFTools.aim_select"><code>VCFTools.aim_select</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">aim_select(vcffile::AbstractString, sampleID_to_population::Dict{String, String}, 
    [maf_threshold::Float64=0.01])</code></pre><p>Ranks SNPs by their ancestry information content. All people should  be assigned an ancestry origin and be fully typed. P-values are computed  via a likelihood ratio heterogeneity test. Sex chromosome is ignored (for now). </p><p><strong>Inputs</strong></p><ul><li><code>vcffile</code>: VCF file, ending with <code>.vcf</code> or <code>.vcf.gz</code></li><li><code>sampleID_to_population</code>: A dictionary mapping every sample ID to a population   origin. </li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>maf_threshold</code>: Minimum minor allele frequency for SNPs considered (default 0.01)</li></ul><p><strong>Note:</strong></p><p>Method is adapted from MendelAimSelection.jl https://github.com/OpenMendel/MendelAimSelection.jl</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/OpenMendel/VCFTools.jl/blob/ef6648bb35d855d22af6b26eedd7ad01d7c556bb/src/aim_select.jl#L1-L20">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../aim/">« Ancestry informative markers</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.25 on <span class="colophon-date" title="Tuesday 11 July 2023 02:29">Tuesday 11 July 2023</span>. Using Julia version 1.9.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
