<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Convert · VCFTools</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">VCFTools</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../summaryinfo/">VCF Summary</a></li><li><a class="tocitem" href="../filter/">Filter</a></li><li class="is-active"><a class="tocitem" href>Convert</a><ul class="internal"><li><a class="tocitem" href="#Example-VCF-file"><span>Example VCF file</span></a></li><li><a class="tocitem" href="#Genetic-model"><span>Genetic model</span></a></li><li><a class="tocitem" href="#Convert-GT-data-into-a-numeric-genotype-matrix"><span>Convert GT data into a numeric genotype matrix</span></a></li><li><a class="tocitem" href="#Convert-GT-data-into-haplotypes-panels"><span>Convert GT data into haplotypes panels</span></a></li><li><a class="tocitem" href="#Convert-DS-data-into-dosages"><span>Convert DS data into dosages</span></a></li><li><a class="tocitem" href="#Extract-data-marker-by-maker-or-window-by-window"><span>Extract data marker-by-maker or window-by-window</span></a></li><li><a class="tocitem" href="#Sample-ID,-Chromosome,-SNP-position,-REF/ALT-alleles"><span>Sample ID, Chromosome, SNP position, REF/ALT alleles</span></a></li></ul></li><li><a class="tocitem" href="../write/">Write</a></li><li><a class="tocitem" href="../conformgt/">Match markers in two VCF</a></li><li><a class="tocitem" href="../grm/">Genetic Relationship Matrix</a></li><li><a class="tocitem" href="../aim/">Ancestry informative markers</a></li><li><a class="tocitem" href="../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Convert</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Convert</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/VCFTools.jl/blob/master/docs/src/man/convert.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Convert-genotype/haplotype/dosage-to-numeric-arrays"><a class="docs-heading-anchor" href="#Convert-genotype/haplotype/dosage-to-numeric-arrays">Convert genotype/haplotype/dosage to numeric arrays</a><a id="Convert-genotype/haplotype/dosage-to-numeric-arrays-1"></a><a class="docs-heading-anchor-permalink" href="#Convert-genotype/haplotype/dosage-to-numeric-arrays" title="Permalink"></a></h1><p>Most often we need to convert genetic data to numeric arrays for statistical analysis.</p><h2 id="Example-VCF-file"><a class="docs-heading-anchor" href="#Example-VCF-file">Example VCF file</a><a id="Example-VCF-file-1"></a><a class="docs-heading-anchor-permalink" href="#Example-VCF-file" title="Permalink"></a></h2><p>We need an example VCF file for demonstation. You can manually download it from <a href="http://faculty.washington.edu/browning/beagle/test.08Jun17.d8b.vcf.gz">link</a> (877KB) and put the file in your current working directory. Or, within Julia,</p><pre><code class="language-julia hljs">isfile(&quot;test.08Jun17.d8b.vcf.gz&quot;) || download(&quot;http://faculty.washington.edu/browning/beagle/test.08Jun17.d8b.vcf.gz&quot;, 
    joinpath(pwd(), &quot;test.08Jun17.d8b.vcf.gz&quot;))
stat(&quot;test.08Jun17.d8b.vcf.gz&quot;)</code></pre><pre><code class="nohighlight hljs">StatStruct(mode=0o100644, size=876514)</code></pre><p>The first 5 markers in this VCF file are</p><pre><code class="language-julia hljs">using VCFTools

fh = openvcf(&quot;test.08Jun17.d8b.vcf.gz&quot;, &quot;r&quot;)
for l in 1:35
    println(readline(fh))
end
close(fh)</code></pre><pre><code class="nohighlight hljs">##fileformat=VCFv4.1
##INFO=&lt;ID=LDAF,Number=1,Type=Float,Description=&quot;MLE Allele Frequency Accounting for LD&quot;&gt;
##INFO=&lt;ID=AVGPOST,Number=1,Type=Float,Description=&quot;Average posterior probability from MaCH/Thunder&quot;&gt;
##INFO=&lt;ID=RSQ,Number=1,Type=Float,Description=&quot;Genotype imputation quality from MaCH/Thunder&quot;&gt;
##INFO=&lt;ID=ERATE,Number=1,Type=Float,Description=&quot;Per-marker Mutation rate from MaCH/Thunder&quot;&gt;
##INFO=&lt;ID=THETA,Number=1,Type=Float,Description=&quot;Per-marker Transition rate from MaCH/Thunder&quot;&gt;
##INFO=&lt;ID=CIEND,Number=2,Type=Integer,Description=&quot;Confidence interval around END for imprecise variants&quot;&gt;
##INFO=&lt;ID=CIPOS,Number=2,Type=Integer,Description=&quot;Confidence interval around POS for imprecise variants&quot;&gt;
##INFO=&lt;ID=END,Number=1,Type=Integer,Description=&quot;End position of the variant described in this record&quot;&gt;
##INFO=&lt;ID=HOMLEN,Number=.,Type=Integer,Description=&quot;Length of base pair identical micro-homology at event breakpoints&quot;&gt;
##INFO=&lt;ID=HOMSEQ,Number=.,Type=String,Description=&quot;Sequence of base pair identical micro-homology at event breakpoints&quot;&gt;
##INFO=&lt;ID=SVLEN,Number=1,Type=Integer,Description=&quot;Difference in length between REF and ALT alleles&quot;&gt;
##INFO=&lt;ID=SVTYPE,Number=1,Type=String,Description=&quot;Type of structural variant&quot;&gt;
##INFO=&lt;ID=AC,Number=.,Type=Integer,Description=&quot;Alternate Allele Count&quot;&gt;
##INFO=&lt;ID=AN,Number=1,Type=Integer,Description=&quot;Total Allele Count&quot;&gt;
##ALT=&lt;ID=DEL,Description=&quot;Deletion&quot;&gt;
##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;
##FORMAT=&lt;ID=DS,Number=1,Type=Float,Description=&quot;Genotype dosage from MaCH/Thunder&quot;&gt;
##FORMAT=&lt;ID=GL,Number=.,Type=Float,Description=&quot;Genotype Likelihoods&quot;&gt;
##INFO=&lt;ID=AA,Number=1,Type=String,Description=&quot;Ancestral Allele, ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/pilot_data/technical/reference/ancestral_alignments/README&quot;&gt;
##INFO=&lt;ID=AF,Number=1,Type=Float,Description=&quot;Global Allele Frequency based on AC/AN&quot;&gt;
##INFO=&lt;ID=AMR_AF,Number=1,Type=Float,Description=&quot;Allele Frequency for samples from AMR based on AC/AN&quot;&gt;
##INFO=&lt;ID=ASN_AF,Number=1,Type=Float,Description=&quot;Allele Frequency for samples from ASN based on AC/AN&quot;&gt;
##INFO=&lt;ID=AFR_AF,Number=1,Type=Float,Description=&quot;Allele Frequency for samples from AFR based on AC/AN&quot;&gt;
##INFO=&lt;ID=EUR_AF,Number=1,Type=Float,Description=&quot;Allele Frequency for samples from EUR based on AC/AN&quot;&gt;
##INFO=&lt;ID=VT,Number=1,Type=String,Description=&quot;indicates what type of variant the line represents&quot;&gt;
##INFO=&lt;ID=SNPSOURCE,Number=.,Type=String,Description=&quot;indicates if a snp was called when analysing the low coverage or exome alignment data&quot;&gt;
##reference=GRCh37
##reference=GRCh37
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	HG00096	HG00097	HG00099	HG00100	HG00101	HG00102	HG00103	HG00104	HG00106	HG00108	HG00109	HG00110	HG00111	HG00112	HG00113	HG00114	HG00116	HG00117	HG00118	HG00119	HG00120	HG00121	HG00122	HG00123	HG00124	HG00125	HG00126	HG00127	HG00128	HG00129	HG00130	HG00131	HG00133	HG00134	HG00135	HG00136	HG00137	HG00138	HG00139	HG00140	HG00141	HG00142	HG00143	HG00146	HG00148	HG00149	HG00150	HG00151	HG00152	HG00154	HG00155	HG00156	HG00158	HG00159	HG00160	HG00171	HG00173	HG00174	HG00176	HG00177	HG00178	HG00179	HG00180	HG00182	HG00183	HG00185	HG00186	HG00187	HG00188	HG00189	HG00190	HG00231	HG00232	HG00233	HG00234	HG00235	HG00236	HG00237	HG00238	HG00239	HG00240	HG00242	HG00243	HG00244	HG00245	HG00246	HG00247	HG00249	HG00250	HG00251	HG00252	HG00253	HG00254	HG00255	HG00256	HG00257	HG00258	HG00259	HG00260	HG00261	HG00262	HG00263	HG00264	HG00265	HG00266	HG00267	HG00268	HG00269	HG00270	HG00271	HG00272	HG00273	HG00274	HG00275	HG00276	HG00277	HG00278	HG00280	HG00281	HG00282	HG00284	HG00285	HG00306	HG00309	HG00310	HG00311	HG00312	HG00313	HG00315	HG00318	HG00319	HG00320	HG00321	HG00323	HG00324	HG00325	HG00326	HG00327	HG00328	HG00329	HG00330	HG00331	HG00332	HG00334	HG00335	HG00336	HG00337	HG00338	HG00339	HG00341	HG00342	HG00343	HG00344	HG00345	HG00346	HG00349	HG00350	HG00351	HG00353	HG00355	HG00356	HG00357	HG00358	HG00359	HG00360	HG00361	HG00362	HG00364	HG00366	HG00367	HG00369	HG00372	HG00373	HG00375	HG00376	HG00377	HG00378	HG00381	HG00382	HG00383	HG00384	HG00403	HG00404	HG00406	HG00407	HG00418	HG00419	HG00421	HG00422	HG00427	HG00428
22	20000086	rs138720731	T	C	100	PASS	AC=7;RSQ=0.8454;AVGPOST=0.9983;AA=T;AN=2184;LDAF=0.0040;THETA=0.0001;VT=SNP;SNPSOURCE=LOWCOV;ERATE=0.0003;AF=0.0032;AFR_AF=0.01	GT:DS:GL	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.04,-1.05,-5.00	0/0:0.000:-0.07,-0.85,-5.00	0/0:0.000:-0.03,-1.18,-5.00	0/0:0.000:-0.06,-0.87,-5.00	0/0:0.000:-0.03,-1.14,-5.00	0/0:0.000:-0.06,-0.90,-5.00	0/0:0.000:-0.23,-0.45,-1.28	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.11,-0.65,-4.40	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.06,-0.91,-5.00	0/0:0.000:-0.18,-0.47,-2.54	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.06,-0.90,-5.00	0/0:0.000:-0.01,-1.74,-5.00	0/0:0.000:-0.00,-3.66,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.00,-2.53,-5.00	0/0:0.000:-0.09,-0.73,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.00,-3.11,-5.00	0/0:0.000:-0.06,-0.89,-5.00	0/0:0.000:-0.09,-0.71,-4.10	0/0:0.000:-0.11,-0.65,-4.40	0/0:0.000:-0.18,-0.47,-2.34	0/0:0.000:-0.22,-0.45,-1.32	0/0:0.000:-0.02,-1.29,-5.00	0/0:0.000:-0.03,-1.15,-5.00	0/0:0.000:-0.02,-1.45,-5.00	0/0:0.000:-0.00,-3.34,-5.00	0/0:0.000:-0.12,-0.61,-3.19	0/0:0.000:-0.11,-0.67,-4.40	0/0:0.000:-0.05,-0.99,-5.00	0/0:0.000:-0.18,-0.48,-2.15	0/0:0.000:-0.01,-1.47,-5.00	0/0:0.000:-0.10,-0.67,-3.62	0/0:0.000:-0.03,-1.14,-5.00	0/0:0.000:-0.09,-0.73,-4.40	0/0:0.000:-0.07,-0.84,-4.40	0/0:0.000:-0.18,-0.48,-2.46	0/0:0.000:-0.0292813,-1.18575,-5	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.01,-1.67,-5.00	0/0:0.000:-0.18,-0.47,-2.40	0/0:0.000:-0.03,-1.25,-5.00	0/0:0.000:-0.11,-0.66,-3.44	0/0:0.000:-0.09,-0.73,-4.70	0/0:0.000:-0.0418663,-1.03687,-4.39794	0/0:0.000:-0.08,-0.79,-3.14	0/0:0.000:-0.00,-2.30,-5.00	0/0:0.000:-0.00,-2.54,-5.00	0/0:0.000:-0.03,-1.21,-5.00	0/0:0.000:-0.06,-0.86,-5.00	0/0:0.000:-0.09,-0.71,-4.70	0/0:0.000:-0.01,-1.49,-5.00	0/0:0.000:-0.01,-1.88,-5.00	0/0:0.000:-0.09,-0.71,-4.70	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.10,-0.67,-4.40	0/0:0.000:-0.01,-1.51,-5.00	0/0:0.000:-0.02,-1.40,-5.00	0/0:0.000:-0.03,-1.17,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.01,-1.48,-5.00	0/0:0.000:-0.00,-2.02,-5.00	0/0:0.000:-0.03,-1.18,-5.00	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.03,-1.17,-5.00	0/0:0.050:-0.18,-0.47,-2.73	0/0:0.000:-0.17,-0.49,-2.97	0/0:0.000:-0.10,-0.68,-4.40	0/0:0.000:-0.05,-0.99,-5.00	0/0:0.000:-0.12,-0.62,-3.38	0/0:0.000:-0.00,-2.06,-5.00	0/0:0.000:-0.16,-0.51,-2.66	0/0:0.000:-0.11,-0.64,-4.22	0/0:0.000:-0.03,-1.22,-5.00	0/0:0.000:-0.01,-1.64,-5.00	0/0:0.000:-0.00,-2.85,-5.00	0/0:0.000:-0.02,-1.38,-5.00	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.0311436,-1.15989,-5	0/0:0.000:-0.36,-0.42,-0.73	0/0:0.000:-0.01,-1.88,-5.00	0/0:0.000:-0.05,-0.92,-5.00	0/0:0.000:-0.03,-1.16,-5.00	0/0:0.000:-0.04,-1.04,-5.00	0/0:0.000:-0.13,-0.59,-5.00	0/0:0.000:-0.02,-1.36,-5.00	0/0:0.000:-0.16,-0.51,-2.36	0/0:0.000:-0.02,-1.31,-5.00	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.03,-1.17,-5.00	0/0:0.000:-0.00,-4.40,-5.00	0/0:0.000:-0.03,-1.16,-5.00	0/0:0.000:-0.09,-0.73,-3.70	0/0:0.000:-0.19,-0.47,-1.77	0/0:0.000:-0.00,-3.32,-5.00	0/0:0.000:-0.17,-0.51,-2.00	0/0:0.000:-0.00,-2.17,-5.00	0/0:0.000:-0.00,-2.91,-5.00	0/0:0.000:-0.10,-0.71,-4.10	0/0:0.000:-0.03,-1.12,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.01,-1.67,-5.00	0/0:0.000:-0.00,-2.09,-5.00	0/0:0.000:-0.04,-1.09,-5.00	0/0:0.000:-0.01,-1.48,-5.00	0/0:0.000:-0.02,-1.41,-5.00	0/0:0.000:-0.10,-0.69,-3.80	0/0:0.000:-0.01,-1.54,-5.00	0/0:0.000:-0.03,-1.16,-5.00	0/0:0.000:-0.09,-0.73,-4.70	0/0:0.000:-0.09,-0.74,-4.70	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.05,-0.97,-5.00	0/0:0.000:-0.08,-0.78,-5.00	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.10,-0.67,-4.40	0/0:0.000:-0.01,-1.71,-5.00	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.02,-1.26,-5.00	0/0:0.000:-0.04,-1.10,-5.00	0/0:0.000:-0.02,-1.27,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.01,-1.47,-5.00	0/0:0.000:-0.00,-2.00,-5.00	0/0:0.000:-0.10,-0.67,-4.22	0/0:0.050:-0.18,-0.47,-2.34	0/0:0.000:-0.05,-1.00,-5.00	0/0:0.000:-0.11,-0.65,-3.85	0/0:0.000:-0.10,-0.68,-4.70	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.08,-0.76,-5.00	0/0:0.000:-0.19,-0.47,-2.14	0/0:0.000:-0.00,-1.99,-5.00	0/0:0.000:-0.18,-0.47,-2.46	0/0:0.000:-0.09,-0.74,-4.40	0/0:0.450:-0.05,-0.94,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.10,-0.69,-4.70	0/0:0.000:-0.01,-1.50,-5.00	0/0:0.000:-0.18,-0.47,-2.34	0/0:0.000:-0.03,-1.17,-5.00	0/0:0.000:-0.06,-0.88,-5.00	0/0:0.000:-0.02,-1.41,-5.00	0/0:0.000:-0.06,-0.88,-5.00	0/0:0.000:-0.18,-0.47,-1.95	0/0:0.000:-0.19,-0.46,-2.17	0/0:0.000:-0.03,-1.13,-5.00	0/0:0.000:-0.03,-1.18,-5.00	0/0:0.000:-0.18,-0.48,-2.23	0/0:0.000:-0.23,-0.45,-1.31	0/0:0.000:-0.11,-0.64,-3.92	0/0:0.000:-0.03,-1.17,-5.00	0/0:0.000:-0.03,-1.22,-5.00	0/0:0.000:-0.11,-0.66,-4.22	0/0:0.000:-0.12,-0.61,-2.38	0/0:0.000:-0.03,-1.22,-5.00	0/0:0.000:-0.40,-0.45,-0.60	0/0:0.000:-0.00,-2.98,-5.00	0/0:0.000:-0.13,-0.59,-2.09	0/0:0.000:-0.02,-1.37,-5.00	0/0:0.000:-0.477139,-0.477113,-0.477113	0/0:0.000:-0.04,-1.10,-5.00	0/0:0.000:-0.03,-1.23,-5.00	0/0:0.000:-0.01,-1.51,-5.00	0/0:0.000:-0.01,-1.67,-5.00	0/0:0.000:-0.08,-0.75,-4.40	0/0:0.000:-0.03,-1.23,-5.00	0/0:0.000:-0.10,-0.69,-4.40	0/0:0.000:-0.12,-0.63,-3.92	0/0:0.000:-0.01,-1.74,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.19,-0.46,-2.60	0/0:0.000:-0.19,-0.46,-2.62	0/0:0.000:-0.11,-0.65,-4.70	0/0:0.000:-0.11,-0.66,-4.70	0/0:0.050:-0.18,-0.49,-2.04	0/0:0.050:-0.10,-0.67,-4.40	0/0:0.000:-0.01,-1.62,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.23,-0.41,-1.51	0/0:0.000:-0.18,-0.48,-2.18	0/0:0.000:-0.03,-1.17,-5.00	0/0:0.000:-0.10,-0.68,-4.10	0/0:0.000:-0.03,-1.24,-5.00	0/0:0.000:-0.18,-0.48,-2.14
22	20000146	rs73387790	G	A	100	PASS	LDAF=0.0169;RSQ=0.9482;THETA=0.0004;AA=G;AN=2184;AVGPOST=0.9972;VT=SNP;SNPSOURCE=LOWCOV;AC=36;ERATE=0.0003;AF=0.02;AFR_AF=0.07;EUR_AF=0.0013	GT:DS:GL	0/0:0.000:-0.00,-2.68,-5.00	0/0:0.000:-0.07,-0.82,-5.00	0/0:0.000:-0.13,-0.60,-3.05	0/0:0.000:-0.03,-1.24,-5.00	0/0:0.000:-0.18,-0.47,-3.08	0/0:0.000:-0.06,-0.90,-5.00	0/0:0.000:-0.01,-1.48,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.16,-0.51,-2.63	0/0:0.000:-0.01,-1.76,-5.00	0/0:0.000:-0.10,-0.67,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.11,-0.66,-4.40	0/0:0.000:-0.10,-0.69,-4.70	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.18,-0.47,-2.30	0/0:0.000:-0.00,-2.80,-5.00	0/0:0.000:-0.00,-2.02,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.01,-1.49,-5.00	0/0:0.000:-0.01,-1.76,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.02,-1.40,-5.00	0/0:0.000:-0.00,-2.09,-5.00	0/0:0.000:-0.10,-0.70,-4.10	0/0:0.000:-0.22,-0.46,-1.27	0/0:0.000:-0.18,-0.48,-2.39	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.19,-0.47,-2.27	0/0:0.000:-0.07,-0.85,-5.00	0/0:0.000:-0.00,-2.53,-5.00	0/0:0.000:-0.00,-2.83,-5.00	0/0:0.000:-0.22,-0.46,-1.24	0/0:0.000:-0.19,-0.46,-2.27	0/0:0.000:-0.10,-0.68,-4.40	0/0:0.000:-0.09,-0.73,-4.22	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.15,-0.55,-2.64	0/0:0.000:-0.05,-0.97,-5.00	0/0:0.000:-0.08,-0.76,-4.70	0/0:0.000:-0.01,-1.49,-5.00	0/0:0.000:-0.06,-0.86,-5.00	0/0:0.000:-0.029681,-1.18006,-5	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.18,-0.48,-2.33	0/0:0.000:-0.10,-0.70,-4.70	0/0:0.000:-0.00,-2.04,-5.00	0/0:0.000:-0.03,-1.24,-5.00	0/0:0.000:-0.10,-0.69,-4.40	0/0:0.000:-0.0843997,-0.755377,-3.00877	0/0:0.000:-0.21,-0.47,-1.33	0/0:0.000:-0.00,-3.22,-5.00	0/0:0.000:-0.00,-3.70,-5.00	0/0:0.000:-0.05,-0.96,-5.00	0/0:0.000:-0.03,-1.21,-5.00	0/0:0.000:-0.01,-1.47,-5.00	0/0:0.000:-0.01,-1.86,-5.00	0/0:0.000:-0.05,-0.97,-5.00	0/0:0.000:-0.06,-0.90,-5.00	0/0:0.000:-0.01,-1.51,-5.00	0/0:0.000:-0.03,-1.25,-5.00	0/0:0.000:-0.01,-1.92,-5.00	0/0:0.000:-0.00,-2.58,-5.00	0/0:0.000:-0.06,-0.89,-5.00	0/0:0.000:-0.05,-1.00,-5.00	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.01,-1.72,-5.00	0/0:0.000:-0.00,-2.25,-5.00	0/0:0.000:-0.02,-1.43,-5.00	0/0:0.000:-0.18,-0.48,-2.03	0/0:0.000:-0.18,-0.47,-2.72	0/0:0.000:-0.09,-0.72,-4.70	0/0:0.000:-0.18,-0.47,-2.42	0/0:0.000:-0.19,-0.46,-2.20	0/0:0.000:-0.24,-0.44,-1.19	0/0:0.000:-0.01,-1.75,-5.00	0/0:0.000:-0.01,-1.48,-5.00	0/0:0.000:-0.19,-0.46,-2.27	0/0:0.000:-0.05,-0.99,-5.00	0/0:0.000:-0.06,-0.87,-5.00	0/0:0.000:-0.00,-3.40,-5.00	0/0:0.000:-0.10,-0.68,-4.70	0/0:0.000:-0.01,-1.72,-5.00	0/0:0.000:-0.0148341,-1.47392,-5	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.03,-1.16,-5.00	0/0:0.000:-0.04,-1.03,-5.00	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.09,-0.73,-4.40	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.02,-1.26,-5.00	0/0:0.000:-0.03,-1.25,-5.00	0/0:0.000:-0.19,-0.45,-2.12	0/0:0.000:-0.01,-1.50,-5.00	0/0:0.000:-0.05,-0.96,-5.00	0/0:0.000:-0.00,-4.70,-5.00	0/0:0.000:-0.13,-0.58,-3.06	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.17,-0.51,-2.09	0/0:0.000:-0.00,-2.29,-5.00	0/0:0.000:-0.38,-0.43,-0.67	0/0:0.000:-0.00,-2.81,-5.00	0/0:0.000:-0.00,-3.25,-5.00	0/0:0.000:-0.22,-0.46,-1.26	0/0:0.000:-0.03,-1.17,-5.00	0/0:0.000:-0.03,-1.21,-5.00	0/0:0.000:-0.01,-1.50,-5.00	0/0:0.000:-0.01,-1.54,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.00,-2.04,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.10,-0.68,-3.70	0/0:0.000:-0.09,-0.72,-5.00	0/0:0.000:-0.01,-1.77,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.01,-1.51,-5.00	0/0:0.000:-0.16,-0.51,-2.74	0/0:0.000:-0.10,-0.69,-4.40	0/0:0.000:-0.18,-0.48,-2.26	0/0:0.000:-0.18,-0.48,-2.37	0/0:0.000:-0.18,-0.48,-2.27	0/0:0.000:-0.00,-2.58,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.00,-2.34,-5.00	0/0:0.000:-0.03,-1.23,-5.00	0/0:0.000:-0.10,-0.69,-4.40	0/0:0.000:-0.08,-0.78,-4.70	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.03,-1.22,-5.00	0/0:0.000:-0.18,-0.48,-2.36	0/0:0.000:-0.01,-1.53,-5.00	0/0:0.000:-0.18,-0.48,-2.25	0/0:0.000:-0.10,-0.68,-4.70	0/0:0.000:-0.09,-0.73,-5.00	0/0:0.000:-0.02,-1.41,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.03,-1.22,-5.00	0/0:0.000:-0.18,-0.47,-2.41	0/0:0.000:-0.09,-0.73,-4.40	0/0:0.000:-0.00,-2.00,-5.00	0/0:0.000:-0.19,-0.46,-2.19	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.05,-0.98,-5.00	0/0:0.000:-0.18,-0.47,-2.31	0/0:0.000:-0.09,-0.73,-4.10	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.10,-0.67,-3.66	0/0:0.000:-0.12,-0.63,-4.70	0/0:0.000:-0.16,-0.51,-2.28	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.01,-1.75,-5.00	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.10,-0.68,-4.22	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.12,-0.62,-3.08	0/0:0.000:-0.19,-0.45,-2.25	0/0:0.000:-0.01,-1.77,-5.00	0/0:0.000:-0.13,-0.60,-2.15	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.04,-1.05,-5.00	0/0:0.000:-0.19,-0.46,-2.30	0/0:0.000:-0.19,-0.46,-2.28	0/0:0.000:-0.01,-1.48,-5.00	0/0:0.000:-0.190265,-0.457324,-2.2321	0/0:0.000:-0.23,-0.46,-1.21	0/0:0.000:-0.01,-1.50,-5.00	0/0:0.000:-0.00,-2.44,-5.00	0/0:0.000:-0.01,-1.73,-5.00	0/0:0.000:-0.01,-1.53,-5.00	0/0:0.000:-0.05,-0.96,-5.00	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.19,-0.46,-2.62	0/0:0.000:-0.00,-2.25,-5.00	0/0:0.000:-0.19,-0.46,-2.23	0/0:0.000:-0.11,-0.65,-4.70	0/0:0.000:-0.18,-0.46,-2.68	0/0:0.000:-0.11,-0.66,-4.40	0/0:0.000:-0.19,-0.46,-2.43	0/0:0.000:-0.18,-0.49,-2.00	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.20,-0.45,-2.05	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.03,-1.18,-5.00	0/0:0.000:-0.10,-0.68,-4.00	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.10,-0.69,-4.22	0/0:0.000:-0.11,-0.65,-4.40	0/0:0.000:-0.11,-0.67,-3.85
22	20000199	rs183293480	A	C	100	PASS	LDAF=0.0009;THETA=0.0004;AN=2184;AVGPOST=0.9990;VT=SNP;AA=A;RSQ=0.6274;SNPSOURCE=LOWCOV;AC=1;ERATE=0.0003;AF=0.0005;EUR_AF=0.0013	GT:DS:GL	0/0:0.000:-0.00,-2.04,-5.00	0/0:0.000:-0.07,-0.82,-3.47	0/0:0.000:-0.07,-0.83,-5.00	0/0:0.000:-0.03,-1.12,-5.00	0/0:0.000:-0.11,-0.64,-4.10	0/0:0.000:-0.12,-0.62,-3.85	0/0:0.000:-0.01,-1.47,-5.00	0/0:0.000:-0.01,-1.54,-5.00	0/0:0.000:-0.10,-0.70,-4.70	0/0:0.000:-0.03,-1.18,-5.00	0/0:0.000:-0.16,-0.50,-3.30	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.10,-0.70,-5.00	0/0:0.000:-0.19,-0.46,-2.46	0/0:0.000:-0.16,-0.51,-2.67	0/0:0.000:-0.00,-2.57,-5.00	0/0:0.000:-0.00,-2.85,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.00,-2.55,-5.00	0/0:0.000:-0.00,-2.02,-5.00	0/0:0.050:-0.48,-0.48,-0.48	0/0:0.000:-0.23,-0.46,-1.24	0/0:0.000:-0.02,-1.45,-5.00	0/0:0.000:-0.10,-0.68,-4.40	0/0:0.000:-0.06,-0.88,-5.00	0/0:0.000:-0.13,-0.61,-2.23	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.26,-0.43,-1.11	0/0:0.000:-0.04,-1.01,-5.00	0/0:0.000:-0.12,-0.62,-3.28	0/0:0.000:-0.00,-2.27,-5.00	0/0:0.000:-0.00,-2.42,-5.00	0/0:0.000:-0.04,-1.08,-5.00	0/0:0.000:-0.06,-0.88,-5.00	0/0:0.000:-0.04,-1.03,-5.00	0/0:0.000:-0.10,-0.70,-4.10	0/0:0.000:-0.18,-0.48,-2.22	0/0:0.000:-0.02,-1.41,-5.00	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.14,-0.57,-2.12	0/0:0.000:-0.10,-0.69,-4.70	0/0:0.000:-0.12,-0.62,-2.28	0/0:0.000:-0.0149779,-1.4698,-5	0/0:0.000:-0.10,-0.70,-4.70	0/0:0.000:-0.06,-0.89,-5.00	0/0:0.000:-0.39,-0.23,-2.25	0/0:0.000:-0.00,-3.40,-5.00	0/0:0.000:-0.07,-0.86,-4.40	0/0:0.000:-0.18,-0.48,-2.35	0/0:0.000:-0.00967891,-1.65679,-5	0/0:0.000:-0.22,-0.46,-1.24	0/0:0.000:-0.00,-2.27,-5.00	0/0:0.000:-0.00,-3.12,-5.00	0/0:0.000:-0.01,-1.73,-5.00	0/0:0.000:-0.02,-1.39,-5.00	0/0:0.000:-0.10,-0.70,-4.70	0/0:0.000:-0.03,-1.23,-5.00	0/0:0.000:-0.09,-0.72,-4.10	0/0:0.000:-0.10,-0.71,-4.70	0/0:0.000:-0.02,-1.45,-5.00	0/0:0.000:-0.04,-1.11,-5.00	0/0:0.000:-0.03,-1.14,-5.00	0/0:0.000:-0.00,-2.42,-5.00	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.11,-0.65,-3.85	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.01,-1.72,-5.00	0/0:0.000:-0.00,-2.08,-5.00	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.16,-0.52,-2.11	0/0:0.000:-0.10,-0.69,-4.70	0/0:0.000:-0.44,-0.46,-0.54	0/0:0.000:-0.07,-0.83,-5.00	0/0:0.000:-0.16,-0.51,-2.42	0/0:0.000:-0.18,-0.48,-2.30	0/0:0.000:-0.19,-0.46,-2.27	0/0:0.000:-0.06,-0.89,-5.00	0/0:0.000:-0.06,-0.88,-5.00	0/0:0.000:-0.00,-3.15,-5.00	0/0:0.000:-0.12,-0.61,-4.10	0/0:0.000:-0.06,-0.91,-5.00	0/0:0.000:-0.00935928,-1.67121,-5	0/0:0.000:-0.18,-0.48,-2.13	0/0:0.000:-0.07,-0.85,-5.00	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.03,-1.15,-5.00	0/0:0.000:-0.09,-0.71,-4.70	0/0:0.000:-0.19,-0.46,-2.52	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.09,-0.73,-4.10	0/0:0.000:-0.11,-0.64,-4.40	0/0:0.000:-0.03,-1.21,-5.00	0/0:0.000:-0.19,-0.47,-1.83	0/0:0.000:-0.00,-2.76,-5.00	0/0:0.000:-0.09,-0.73,-3.92	0/0:0.000:-0.15,-0.55,-2.43	0/0:0.000:-0.18,-0.48,-1.89	0/0:0.000:-0.00,-3.18,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.02,-1.43,-5.00	0/0:0.000:-0.22,-0.40,-5.00	0/0:0.000:-0.07,-0.84,-4.70	0/0:0.000:-0.03,-1.22,-5.00	0/0:0.000:-0.10,-0.69,-3.85	0/0:0.000:-0.01,-1.84,-5.00	0/0:0.000:-0.00,-3.07,-5.00	0/0:0.000:-0.10,-0.69,-3.85	0/0:0.000:-0.06,-0.91,-5.00	0/0:0.000:-0.10,-0.68,-4.70	0/0:0.000:-0.10,-0.69,-3.80	0/0:0.000:-0.48,-0.18,-4.10	0/0:0.000:-0.02,-1.47,-5.00	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.03,-1.22,-5.00	0/0:0.000:-0.18,-0.48,-2.34	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.18,-0.48,-2.37	0/0:0.000:-0.10,-0.70,-4.40	0/0:0.000:-0.00,-2.44,-5.00	0/0:0.000:-0.01,-1.78,-5.00	0/0:0.000:-0.00,-2.06,-5.00	0/0:0.000:-0.00,-2.28,-5.00	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.03,-1.24,-5.00	0/0:0.000:-0.01,-1.51,-5.00	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.11,-0.64,-4.70	0/0:0.000:-0.19,-0.46,-2.54	0/0:0.000:-0.02,-1.38,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.02,-1.26,-5.00	0/0:0.000:-0.18,-0.47,-2.47	0/0:0.000:-0.16,-0.51,-2.43	0/0:0.000:-0.01,-1.73,-5.00	0/0:0.000:-0.19,-0.46,-2.79	0/0:0.000:-0.06,-0.91,-5.00	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.18,-0.48,-2.35	0/0:0.000:-0.03,-1.21,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.03,-1.18,-5.00	0/0:0.000:-0.19,-0.46,-2.72	0/0:0.000:-0.02,-1.45,-5.00	0/0:0.000:-0.16,-0.51,-2.71	0/0:0.000:-0.01,-1.80,-5.00	0/0:0.000:-0.10,-0.69,-3.66	0/0:0.000:-0.10,-0.69,-4.22	0/0:0.000:-0.23,-0.44,-1.26	0/0:0.000:-0.19,-0.46,-2.12	0/0:0.000:-0.04,-1.07,-5.00	0/0:0.000:-0.03,-1.23,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.19,-0.46,-2.20	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.11,-0.65,-3.44	0/0:0.000:-0.05,-1.00,-5.00	0/0:0.000:-0.03,-1.17,-5.00	0/0:0.000:-0.01,-1.77,-5.00	0/0:0.000:-0.189687,-0.458221,-2.2426	0/0:0.000:-0.01,-1.79,-5.00	0/0:0.050:-0.24,-0.42,-1.39	0/0:0.000:-0.19,-0.45,-4.70	0/0:0.000:-0.00,-2.15,-5.00	0/0:0.000:-0.05,-0.97,-5.00	0/0:0.000:-0.18,-0.48,-2.23	0/0:0.000:-0.00,-2.11,-5.00	0/0:0.000:-0.11,-0.66,-4.70	0/0:0.000:-0.01,-1.59,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.18,-0.46,-2.63	0/0:0.000:-0.11,-0.66,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.19,-0.46,-2.43	0/0:0.000:-0.19,-0.47,-1.84	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.19,-0.46,-2.21	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.01,-1.82,-5.00	0/0:0.000:-0.05,-0.99,-5.00	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.03,-1.22,-5.00	0/0:0.000:-0.00,-2.26,-5.00	0/0:0.000:-0.10,-0.69,-4.00
22	20000291	rs185807825	G	T	100	PASS	ERATE=0.0005;AVGPOST=0.9983;AA=G;AN=2184;LDAF=0.0015;VT=SNP;SNPSOURCE=LOWCOV;RSQ=0.5564;AC=2;THETA=0.0003;AF=0.0009;ASN_AF=0.0035	GT:DS:GL	0/0:0.000:-0.00,-2.06,-5.00	0/0:0.000:-0.07,-0.83,-5.00	0/0:0.000:-0.02,-1.27,-5.00	0/0:0.000:-0.01,-1.77,-5.00	0/0:0.000:-0.19,-0.45,-2.14	0/0:0.000:-0.11,-0.66,-5.00	0/0:0.000:-0.03,-1.17,-5.00	0/0:0.000:-0.02,-1.33,-5.00	0/0:0.000:-0.18,-0.48,-2.43	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.18,-0.46,-2.76	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.03,-1.15,-5.00	0/0:0.000:-0.07,-0.83,-4.40	0/0:0.000:-0.02,-1.33,-5.00	0/0:0.000:-0.10,-0.68,-4.40	0/0:0.000:-0.00,-2.28,-5.00	0/0:0.000:-0.01,-1.83,-5.00	0/0:0.000:-0.13,-0.61,-2.24	0/0:0.000:-0.00,-3.70,-5.00	0/0:0.000:-0.00,-2.21,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.22,-0.46,-1.25	0/0:0.000:-0.00,-2.81,-5.00	0/0:0.000:-0.02,-1.44,-5.00	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.08,-0.79,-3.21	0/0:0.000:-0.01,-1.72,-5.00	0/0:0.000:-0.21,-0.46,-1.42	0/0:0.000:-0.08,-0.79,-4.40	0/0:0.000:-0.04,-1.06,-5.00	0/0:0.000:-0.02,-1.42,-5.00	0/0:0.000:-0.00,-3.85,-5.00	0/0:0.000:-0.07,-0.83,-5.00	0/0:0.000:-0.01,-1.80,-5.00	0/0:0.000:-0.11,-0.66,-4.40	0/0:0.000:-0.01,-1.48,-5.00	0/0:0.000:-0.01,-1.47,-5.00	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.19,-0.46,-2.54	0/0:0.000:-0.03,-1.22,-5.00	0/0:0.000:-0.05,-1.00,-5.00	0/0:0.000:-0.13,-0.60,-2.19	0/0:0.000:-0.0021071,-2.31515,-5	0/0:0.000:-0.10,-0.68,-4.40	0/0:0.000:-0.10,-0.68,-4.22	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.00,-3.38,-5.00	0/0:0.000:-0.06,-0.90,-5.00	0/0:0.000:-0.20,-0.45,-1.99	0/0:0.000:-0.0193877,-1.35992,-5	0/0:0.000:-0.02,-1.28,-5.00	0/0:0.000:-0.00,-2.46,-5.00	0/0:0.000:-0.00,-2.75,-5.00	0/0:0.000:-0.18,-0.47,-2.36	0/0:0.000:-0.10,-0.68,-4.00	0/0:0.000:-0.02,-1.34,-5.00	0/0:0.000:-0.00,-3.27,-5.00	0/0:0.250:-0.03,-1.19,-5.00	0/0:0.000:-0.07,-0.84,-5.00	0/0:0.000:-0.01,-1.52,-5.00	0/0:0.000:-0.11,-0.65,-3.40	0/0:0.000:-0.10,-0.70,-4.70	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.03,-1.23,-5.00	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.00,-1.99,-5.00	0/0:0.000:-0.01,-1.61,-5.00	0/0:0.000:-0.01,-1.48,-5.00	0/0:0.000:-0.10,-0.69,-4.22	0/0:0.000:-0.01,-1.49,-5.00	0/0:0.000:-0.01,-1.59,-5.00	0/0:0.000:-0.10,-0.69,-3.36	0/0:0.000:-0.01,-1.48,-5.00	0/0:0.000:-0.08,-0.77,-4.10	0/0:0.000:-0.05,-1.00,-5.00	0/0:0.000:-0.10,-0.67,-3.80	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.06,-0.86,-5.00	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.11,-0.66,-4.70	0/0:0.000:-0.00,-2.68,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.01,-1.84,-5.00	0/0:0.000:-0.0293556,-1.18469,-5	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.19,-0.46,-2.37	0/0:0.000:-0.10,-0.68,-4.10	0/0:0.000:-0.05,-0.97,-5.00	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.06,-0.91,-5.00	0/0:0.000:-0.18,-0.48,-2.06	0/0:0.000:-0.06,-0.87,-5.00	0/0:0.000:-0.03,-1.13,-5.00	0/0:0.000:-0.03,-1.24,-5.00	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.00,-3.85,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.07,-0.83,-4.40	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.00,-2.82,-5.00	0/0:0.000:-0.08,-0.76,-3.80	0/0:0.000:-0.00,-3.47,-5.00	0/0:0.000:-0.03,-1.18,-5.00	0/0:0.000:-0.11,-0.65,-3.59	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.02,-1.43,-5.00	0/0:0.000:-0.00,-2.73,-5.00	0/0:0.000:-0.06,-0.86,-4.22	0/0:0.000:-0.14,-0.57,-2.84	0/0:0.000:-0.01,-1.50,-5.00	0/0:0.000:-0.02,-1.45,-5.00	0/0:0.000:-0.01,-1.76,-5.00	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.01,-1.79,-5.00	0/0:0.000:-0.06,-0.90,-5.00	0/0:0.000:-0.01,-1.87,-5.00	0/0:0.000:-0.01,-1.50,-5.00	0/0:0.000:-0.03,-1.21,-5.00	0/0:0.000:-0.08,-0.79,-5.00	0/0:0.000:-0.06,-0.91,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.00,-2.89,-5.00	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.20,-0.44,-2.07	0/0:0.000:-0.00,-2.65,-5.00	0/0:0.000:-0.02,-1.27,-5.00	0/0:0.000:-0.03,-1.21,-5.00	0/0:0.000:-0.01,-1.79,-5.00	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.10,-0.68,-4.70	0/0:0.000:-0.35,-0.41,-0.78	0/0:0.000:-0.09,-0.75,-3.70	0/0:0.000:-0.05,-0.98,-5.00	0/0:0.000:-0.05,-0.97,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.08,-0.77,-5.00	0/0:0.000:-0.04,-1.03,-5.00	0/0:0.000:-0.02,-1.47,-5.00	0/0:0.000:-0.03,-1.16,-5.00	0/0:0.000:-0.01,-1.57,-5.00	0/0:0.000:-0.06,-0.89,-5.00	0/0:0.000:-0.06,-0.90,-5.00	0/0:0.000:-0.00,-2.30,-5.00	0/0:0.000:-0.05,-0.98,-5.00	0/0:0.000:-0.10,-0.69,-4.40	0/0:0.000:-0.01,-1.50,-5.00	0/0:0.000:-0.06,-0.87,-5.00	0/0:0.000:-0.16,-0.51,-2.37	0/0:0.000:-0.19,-0.46,-2.23	0/0:0.050:-0.27,-0.34,-3.22	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.08,-0.77,-4.40	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.05,-0.99,-5.00	0/0:0.000:-0.07,-0.82,-5.00	0/0:0.000:-0.12,-0.61,-2.91	0/0:0.000:-0.00,-1.98,-5.00	0/0:0.000:-0.03,-1.21,-5.00	0/0:0.000:-0.07,-0.82,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.01,-1.49,-5.00	0/0:0.000:-0.12,-0.61,-3.15	0/0:0.000:-0.00,-2.08,-5.00	0/0:0.000:-0.02,-1.33,-5.00	0/0:0.000:-0.01,-1.74,-5.00	0/0:0.050:-0.11227,-0.642561,-4.22185	0/0:0.000:-0.22,-0.46,-1.28	0/0:0.000:-0.10,-0.70,-4.10	0/0:0.000:-0.10,-0.67,-3.62	0/0:0.000:-0.03,-1.18,-5.00	0/0:0.000:-0.03,-1.24,-5.00	0/0:0.000:-0.10,-0.70,-4.10	0/0:0.000:-0.01,-1.80,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.02,-1.47,-5.00	0/0:0.000:-0.06,-0.88,-5.00	0/0:0.000:-0.03,-1.13,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.18,-0.46,-2.69	0/0:0.000:-0.03,-1.13,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.18,-0.48,-2.41	0/0:0.000:-0.18,-0.47,-2.85	0/0:0.050:-0.48,-0.48,-0.48	0/0:0.000:-0.08,-0.76,-4.70	0/0:0.000:-0.10,-0.69,-4.00	0/0:0.000:-0.11,-0.65,-3.62	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.01,-1.70,-5.00	0/0:0.000:-0.00,-2.57,-5.00
22	20000428	rs55902548	G	T	100	PASS	AC=323;AVGPOST=0.9983;AA=G;AN=2184;VT=SNP;RSQ=0.9949;LDAF=0.1473;SNPSOURCE=LOWCOV;ERATE=0.0003;THETA=0.0003;AF=0.15;ASN_AF=0.0017;AMR_AF=0.15;AFR_AF=0.31;EUR_AF=0.15	GT:DS:GL	1/0:1.000:-5.00,0.00,-5.00	0/0:0.000:-0.35,-0.43,-0.73	0/1:1.000:-1.81,-0.01,-2.95	0/0:0.000:-0.01,-1.79,-5.00	0/0:0.000:-0.06,-0.86,-5.00	1/0:1.000:-0.19,-0.46,-2.18	0/0:0.000:-0.10,-0.68,-5.00	0/1:1.000:-4.40,-0.03,-1.12	0/1:1.000:-5.00,-0.69,-0.10	0/0:0.000:-0.10,-0.69,-4.70	0/0:0.000:-0.48,-0.48,-0.48	0/1:1.000:-5.00,-0.01,-1.77	0/0:0.000:-0.18,-0.48,-2.57	0/0:0.000:-0.02,-1.31,-5.00	0/0:0.000:-0.11,-0.65,-4.70	0/0:0.000:-0.10,-0.68,-4.70	0/0:0.000:-0.01,-1.72,-5.00	1/0:1.000:-5.00,0.00,-5.00	1/0:1.000:-1.38,-0.02,-2.61	0/1:1.000:-5.00,-1.40,-0.02	0/0:0.000:-0.00,-2.97,-5.00	0/0:0.000:-0.19,-0.47,-2.15	0/0:0.000:-0.44,-0.46,-0.54	0/0:0.000:-0.00,-2.52,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.01,-1.77,-5.00	0/1:0.750:-0.22,-0.46,-1.26	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.21,-0.46,-1.42	0/0:0.000:-0.19,-0.45,-2.20	0/0:0.000:-0.12,-0.63,-3.36	0/0:0.000:-0.00,-2.54,-5.00	0/0:0.000:-0.00,-2.88,-5.00	1/0:1.000:-2.10,-0.00,-4.00	1/0:1.250:-5.00,-1.62,-0.01	0/1:1.000:-3.06,-0.47,-0.18	0/1:1.000:-5.00,-0.87,-0.06	1/1:2.000:-5.00,-0.84,-0.07	0/0:0.000:-0.05,-0.95,-5.00	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.14,-0.58,-2.17	0/0:0.000:-0.01,-1.77,-5.00	0/0:0.000:-0.06,-0.91,-5.00	1/0:1.000:-5,0,-5	0/0:0.000:-0.05,-0.95,-5.00	1/0:1.000:-4.70,-0.70,-0.10	0/0:0.000:-0.09,-0.72,-4.70	0/0:0.000:-0.01,-1.54,-5.00	1/1:2.000:-5.00,-0.68,-0.10	0/0:0.000:-0.18,-0.48,-2.33	0/0:0.000:-0.00465443,-1.97224,-5	0/0:0.000:-0.48,-0.48,-0.48	1/0:1.000:-5.00,-0.93,-0.05	0/0:0.000:-0.00,-2.82,-5.00	1/1:2.000:-5.00,-1.67,-0.01	0/0:0.000:-0.05,-0.95,-5.00	1/0:1.000:-5.00,-0.87,-0.06	0/0:0.000:-0.01,-1.78,-5.00	0/0:0.000:-0.01,-1.83,-5.00	0/1:1.000:-5.00,-0.00,-3.70	0/1:0.950:-0.15,-0.53,-2.39	0/0:0.000:-0.09,-0.71,-4.70	0/0:0.000:-0.18,-0.48,-2.56	0/0:0.000:-0.00,-2.60,-5.00	0/0:0.000:-0.01,-1.49,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.04,-1.09,-5.00	0/0:0.000:-0.00,-2.40,-5.00	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.01,-1.54,-5.00	0/0:0.000:-0.02,-1.45,-5.00	0/0:0.000:-0.01,-1.84,-5.00	0/1:1.000:-2.39,-0.30,-0.31	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.11,-0.64,-3.44	0/0:0.000:-0.02,-1.42,-5.00	0/0:0.000:-0.01,-1.76,-5.00	0/0:0.000:-0.03,-1.17,-5.00	0/1:1.000:-5.00,-0.03,-1.18	0/0:0.000:-0.03,-1.21,-5.00	0/0:0.000:-0.10,-0.70,-5.00	0/1:1.000:-5.00,0.00,-5.00	0/0:0.000:-0.01,-1.58,-5.00	0/1:1.000:-5.00,0.00,-5.00	1/0:1.000:-5,-5.21375e-05,-3.92082	0/0:0.000:-0.05,-0.92,-5.00	0/0:0.000:-0.02,-1.45,-5.00	0/1:1.000:-5.00,-0.00,-4.22	0/0:0.000:-0.09,-0.73,-4.22	0/0:0.000:-0.00,-2.03,-5.00	0/0:0.000:-0.03,-1.20,-5.00	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.18,-0.48,-2.03	0/0:0.000:-0.01,-1.59,-5.00	0/0:0.000:-0.37,-0.42,-0.71	0/0:0.000:-0.02,-1.45,-5.00	0/0:0.000:-0.00,-3.36,-5.00	1/0:1.000:-3.85,-0.05,-0.94	1/0:1.000:-5.00,-0.84,-0.07	1/0:1.000:-0.06,-0.90,-5.00	0/0:0.000:-0.00,-2.56,-5.00	0/0:0.000:-0.19,-0.47,-1.73	0/0:0.000:-0.06,-0.90,-5.00	0/0:0.000:-0.00,-3.74,-5.00	0/0:0.000:-0.22,-0.46,-1.26	1/0:1.000:-5.00,-0.00,-3.92	0/0:0.000:-0.10,-0.69,-3.85	0/0:0.000:-0.00,-2.76,-5.00	0/0:0.000:-0.01,-1.70,-5.00	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.06,-0.92,-5.00	0/0:0.000:-0.01,-1.73,-5.00	0/0:0.000:-0.05,-1.00,-5.00	1/0:1.000:-2.32,-0.01,-1.58	0/1:0.950:-0.09,-0.73,-4.70	0/1:1.000:-2.25,-0.02,-1.52	0/0:0.000:-0.01,-1.50,-5.00	0/0:0.000:-0.10,-0.70,-4.40	0/0:0.000:-0.05,-0.99,-5.00	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.03,-1.25,-5.00	0/0:0.000:-0.10,-0.67,-4.10	0/1:1.000:-5.00,-0.00,-4.22	0/0:0.000:-0.18,-0.48,-2.02	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.00,-4.70,-5.00	0/0:0.000:-0.03,-1.23,-5.00	0/0:0.000:-0.05,-0.93,-5.00	0/0:0.000:-0.03,-1.20,-5.00	1/0:1.000:-0.37,-0.24,-2.87	0/0:0.000:-0.05,-0.94,-5.00	0/0:0.000:-0.18,-0.48,-2.36	0/0:0.000:-0.18,-0.48,-1.83	0/0:0.000:-0.18,-0.48,-2.23	0/0:0.000:-0.01,-1.78,-5.00	0/0:0.000:-0.06,-0.90,-5.00	0/0:0.000:-0.07,-0.82,-5.00	0/0:0.000:-0.10,-0.70,-4.70	0/0:0.000:-0.01,-1.48,-5.00	0/0:0.000:-0.02,-1.46,-5.00	0/0:0.000:-0.18,-0.48,-2.11	0/0:0.000:-0.10,-0.69,-4.70	0/0:0.000:-0.18,-0.47,-2.84	0/0:0.000:-0.03,-1.21,-5.00	1/0:1.000:-4.40,-0.00,-4.70	0/1:0.800:-0.03,-1.20,-5.00	0/0:0.000:-0.05,-0.98,-5.00	0/0:0.000:-0.10,-0.68,-3.70	0/0:0.000:-0.10,-0.67,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.10,-0.68,-3.59	0/0:0.000:-0.02,-1.26,-5.00	0/0:0.000:-0.48,-0.48,-0.48	0/0:0.000:-0.01,-1.80,-5.00	1/0:1.000:-5.00,-0.66,-0.11	0/0:0.000:-0.12,-0.61,-2.49	1/1:2.000:-5.00,-1.30,-0.02	0/0:0.000:-0.04,-1.07,-5.00	0/0:0.000:-0.18,-0.48,-2.19	0/0:0.000:-0.07,-0.85,-5.00	1/0:1.000:-2.07,-0.04,-1.10	0/1:1.000:-0.09,-0.74,-4.70	0/1:1.000:-5.00,-0.00,-2.47	0/0:0.000:-0.02,-1.26,-5.00	1/0:1.000:-2.25,-0.01,-1.55	0/0:0.000:-0.00,-3.17,-5.00	0/0:0.000:-0.203967,-0.438136,-1.99396	0/0:0.000:-0.12,-0.62,-3.18	0/0:0.000:-0.03,-1.19,-5.00	0/0:0.000:-0.01,-1.66,-5.00	1/0:1.000:-5.00,-0.00,-3.74	1/0:1.000:-1.43,-0.02,-4.70	0/0:0.000:-0.05,-0.96,-5.00	0/0:0.000:-0.09,-0.74,-4.70	0/0:0.000:-0.48,-0.48,-0.48	1/0:1.000:-5.00,-0.18,-0.47	0/0:0.000:-0.17,-0.50,-2.70	0/1:1.000:-5.00,-0.00,-3.80	0/1:1.000:-2.22,-0.01,-1.97	0/1:1.000:-5.00,-0.67,-0.10	0/0:0.000:-0.21,-0.43,-2.01	0/0:0.000:-0.19,-0.47,-1.83	0/0:0.000:-0.10,-0.69,-4.40	0/0:0.000:-0.11,-0.64,-4.10	0/0:0.000:-0.18,-0.47,-2.80	0/0:0.000:-0.01,-1.47,-5.00	0/0:0.000:-0.22,-0.43,-1.61	0/0:0.000:-0.02,-1.47,-5.00	0/0:0.000:-0.01,-1.70,-5.00	0/0:0.000:-0.05,-0.97,-5.00	0/0:0.000:-0.02,-1.28,-5.00</code></pre><p>We&#39;d like to convert the genotype data (GT field) into the matrix of minor alleles.</p><h2 id="Genetic-model"><a class="docs-heading-anchor" href="#Genetic-model">Genetic model</a><a id="Genetic-model-1"></a><a class="docs-heading-anchor-permalink" href="#Genetic-model" title="Permalink"></a></h2><p>There are differnt SNP models. The <em>additive</em> SNP model essentially counts the number of <strong>minor allele</strong> (0, 1 or 2) per genotype. Other SNP models are <em>dominant</em> and <em>recessive</em>, both in terms of the <strong>minor allele</strong>. <code>VCFTools.jl</code> always assume the <code>ALT</code> allele in the VCF file is the minor allele. Thus genotypes are translated to real number according to</p><table><tr><th style="text-align: center">Genotype</th><th style="text-align: center">VCF GT</th><th style="text-align: center"><code>model=:additive</code></th><th style="text-align: center"><code>model=:dominant</code></th><th style="text-align: center"><code>model=:recessive</code></th></tr><tr><td style="text-align: center">ALT, ALT</td><td style="text-align: center">1/1, 1<span>$\vert$</span>1</td><td style="text-align: center">2</td><td style="text-align: center">1</td><td style="text-align: center">1</td></tr><tr><td style="text-align: center">REF, ALT</td><td style="text-align: center">0/1, 0<span>$\vert$</span>1</td><td style="text-align: center">1</td><td style="text-align: center">1</td><td style="text-align: center">0</td></tr><tr><td style="text-align: center">REF, REF</td><td style="text-align: center">0/0, 0<span>$\vert$</span>0</td><td style="text-align: center">0</td><td style="text-align: center">0</td><td style="text-align: center">0</td></tr><tr><td style="text-align: center">missing</td><td style="text-align: center">./., .<span>$\vert$</span>.</td><td style="text-align: center">missing</td><td style="text-align: center">missing</td><td style="text-align: center">missing</td></tr></table><p>To properly record the missing genotypes, VCFTools convert VCF GT data to matrix <code>A</code> where element type of <code>A</code> is either a numeric number, or missing value. In Julia, this means <code>eltype(A) &lt;: Union{Missing, Real}</code> where <code>&lt;:</code> means &quot;is a subtype&quot;. </p><h2 id="Convert-GT-data-into-a-numeric-genotype-matrix"><a class="docs-heading-anchor" href="#Convert-GT-data-into-a-numeric-genotype-matrix">Convert GT data into a numeric genotype matrix</a><a id="Convert-GT-data-into-a-numeric-genotype-matrix-1"></a><a class="docs-heading-anchor-permalink" href="#Convert-GT-data-into-a-numeric-genotype-matrix" title="Permalink"></a></h2><p>Convert GT data in VCF file test.08Jun17.d8b.vcf.gz to a <code>Matrix{Union{Missing, Int8}}</code>. <code>VCFTools.jl</code> will copy the <code>0</code>s and <code>1</code>s of the file directly into <code>A</code> without checking if ALT or REF is the minor allele. </p><pre><code class="language-julia hljs">@time A = convert_gt(Int8, &quot;test.08Jun17.d8b.vcf.gz&quot;; model = :additive, impute = false, center = false, scale = false)</code></pre><pre><code class="nohighlight hljs">  1.760590 seconds (2.06 M allocations: 118.037 MiB, 1.18% gc time, 95.14% compilation time)





191×1356 Matrix{Union{Missing, Int8}}:
 0  0  0  0  1  0  0  0  0  0  0  0  2  …  0  0  0  0  1  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  1  0  0  0  0  0  0  0  2     0  0  0  0  1  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  1  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  1  0  0  0  0  0  0  0  2  …  0  0  0  0  1  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  1  1  0
 0  0  0  0  1  0  0  0  0  0  0  0  2     0  0  0  0  1  0  0  0  0  0  0  0
 0  0  0  0  1  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2  …  0  0  1  0  1  0  0  0  0  1  1  0
 0  0  0  0  1  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  1  1  0
 ⋮              ⋮              ⋮        ⋱     ⋮              ⋮              ⋮
 0  0  0  0  1  0  0  0  0  0  0  0  2     0  0  1  0  0  0  0  0  0  1  1  0
 0  0  0  0  0  0  0  0  0  0  0  0  2  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  2     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  1  0  2  …  0  0  0  0  0  0  0  0  0  0  0  0</code></pre><p>We can also optionally impute missing genotypes according to allele frequency, centers the dosages around 2MAF, and scales the dosages by <code>sqrt(2MAF*(1-MAF))</code>.</p><pre><code class="language-julia hljs">@time A = convert_gt(Float64, &quot;test.08Jun17.d8b.vcf.gz&quot;; model = :additive, impute = true, center = true, scale = true)</code></pre><pre><code class="nohighlight hljs">  0.148316 seconds (523.76 k allocations: 42.592 MiB, 4.85% gc time, 47.83% compilation time)





191×1356 Matrix{Union{Missing, Float64}}:
 0.0  0.0  0.0  0.0   1.41301   0.0  …  0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0   1.41301   0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0   1.41301   0.0  …  0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0   2.36899    2.36899   0.0
 0.0  0.0  0.0  0.0   1.41301   0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0   1.41301   0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0  …  0.0  0.0   2.36899    2.36899   0.0
 0.0  0.0  0.0  0.0   1.41301   0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0   2.36899    2.36899   0.0
 ⋮                              ⋮    ⋱                                  ⋮
 0.0  0.0  0.0  0.0   1.41301   0.0     0.0  0.0   2.36899    2.36899   0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0  …  0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0  …  0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0     0.0  0.0  -0.390016  -0.390016  0.0
 0.0  0.0  0.0  0.0  -0.586138  0.0  …  0.0  0.0  -0.390016  -0.390016  0.0</code></pre><h2 id="Convert-GT-data-into-haplotypes-panels"><a class="docs-heading-anchor" href="#Convert-GT-data-into-haplotypes-panels">Convert GT data into haplotypes panels</a><a id="Convert-GT-data-into-haplotypes-panels-1"></a><a class="docs-heading-anchor-permalink" href="#Convert-GT-data-into-haplotypes-panels" title="Permalink"></a></h2><p>In certain applications such as genotype imputation, the genotypes are <em>phased</em>. In this case, the alleles are separated by a <code>&#39;|&#39;</code> so we can distinguish heterozygote genotypes (<code>1|0</code> vs <code>0|1</code>). Haplotypes can be imported via <code>convert_ht</code></p><pre><code class="language-julia hljs">@time H = convert_ht(Int8, &quot;test.08Jun17.d8b.vcf.gz&quot;)</code></pre><pre><code class="nohighlight hljs">  0.160939 seconds (732.86 k allocations: 52.687 MiB, 5.00% gc time, 50.10% compilation time)





382×1356 Matrix{Int8}:
 0  0  0  0  1  0  0  0  0  0  0  0  1  …  0  0  0  0  1  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  1  0  0  0  0  0  0  0  1  …  0  0  0  0  1  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  1  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  1  0  0  0  0  0  0  0  1  …  0  0  0  0  1  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 ⋮              ⋮              ⋮        ⋱     ⋮              ⋮              ⋮
 0  0  0  0  0  0  0  0  0  0  0  0  1  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1     0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  0  0  1  …  0  0  0  0  0  0  0  0  0  0  0  0
 0  0  0  0  0  0  0  0  0  0  1  0  1     0  0  0  0  0  0  0  0  0  0  0  0</code></pre><p>Note there are 382 rows to represent the haplotypes of 191 samples. The first sample occupies row 1 &amp; 2, the second sample occupies 3&amp;4...etc.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>VCFTools.jl</code> does not check if the allele separators are <code>&#39;|&#39;</code> when you call <code>convert_ht</code>. One must be cautious to run this function only on <em>phased</em> data. </p></div></div><h2 id="Convert-DS-data-into-dosages"><a class="docs-heading-anchor" href="#Convert-DS-data-into-dosages">Convert DS data into dosages</a><a id="Convert-DS-data-into-dosages-1"></a><a class="docs-heading-anchor-permalink" href="#Convert-DS-data-into-dosages" title="Permalink"></a></h2><p>Often data contains dosage information. In this case, matrix values can be any number between 0 and 2. One can similarly import dosage into a numeric matrix.</p><pre><code class="language-julia hljs">@time D = convert_ds(Float64, &quot;test.08Jun17.d8b.vcf.gz&quot;; key=&quot;DS&quot;, impute=false, center=false, scale=false)</code></pre><pre><code class="nohighlight hljs">  0.204783 seconds (1.56 M allocations: 121.663 MiB, 10.37% gc time, 31.82% compilation time)





191×1356 Matrix{Union{Missing, Float64}}:
 0.0   0.0  0.0  0.0   1.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   1.0  0.0  0.0     0.0  0.0  0.0  0.0  0.05  0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   1.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  1.0   1.0  0.0
 0.0   0.0  0.0  0.0   1.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   1.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  1.0   1.0  0.0
 0.0   0.0  0.0  0.0   1.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  1.0   1.0  0.0
 ⋮                          ⋮         ⋱       ⋮                         ⋮
 0.0   0.0  0.0  0.0   1.0  0.0  0.0     0.0  0.0  0.0  0.0  1.0   1.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.05  0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.05  0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.05  0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0     0.0  0.0  0.0  0.0  0.0   0.0  0.0
 0.0   0.0  0.0  0.0   0.0  0.0  0.0  …  0.0  0.0  0.0  0.0  0.0   0.0  0.0</code></pre><h2 id="Extract-data-marker-by-maker-or-window-by-window"><a class="docs-heading-anchor" href="#Extract-data-marker-by-maker-or-window-by-window">Extract data marker-by-maker or window-by-window</a><a id="Extract-data-marker-by-maker-or-window-by-window-1"></a><a class="docs-heading-anchor-permalink" href="#Extract-data-marker-by-maker-or-window-by-window" title="Permalink"></a></h2><p>Large VCF files easily generate numeric arrays that cannot fit into computer memory. Many analyses only need to loop over markers or sets of markers. Previous functions for importing genotypes/haplotypes/dosages have equivalent functions to achieve this:</p><ul><li><strong><code>copy_gt!</code></strong> loops over genotypes</li><li><strong><code>copy_ht!</code></strong> loops over haplotypes</li><li><strong><code>copy_ds!</code></strong> loops over dosages</li></ul><p>For example, to loop over all genotype markers in the VCF file test.08Jun17.d8b.vcf.gz:</p><pre><code class="language-julia hljs">using VariantCallFormat

# initialize VCF reader
people, snps = nsamples(&quot;test.08Jun17.d8b.vcf.gz&quot;), nrecords(&quot;test.08Jun17.d8b.vcf.gz&quot;)
reader = VCF.Reader(openvcf(&quot;test.08Jun17.d8b.vcf.gz&quot;))
# pre-allocate vector for marker data
g = zeros(Union{Missing, Float64}, people)
for j = 1:snps
    copy_gt!(g, reader; model = :additive, impute = true, center = true, scale = true)
    # do statistical anlaysis
end
close(reader)</code></pre><p>To loop over markers in windows of size 25:</p><pre><code class="language-julia hljs"># initialize VCF reader
people, snps = nsamples(&quot;test.08Jun17.d8b.vcf.gz&quot;), nrecords(&quot;test.08Jun17.d8b.vcf.gz&quot;)
reader = VCF.Reader(openvcf(&quot;test.08Jun17.d8b.vcf.gz&quot;))
# pre-allocate matrix for marker data
windowsize = 25
g = zeros(Union{Missing, Float64}, people, windowsize)
nwindows = ceil(Int, snps / windowsize)
for j = 1:nwindows
    copy_gt!(g, reader; model = :additive, impute = true, center = true, scale = true)
    # do statistical anlaysis
end
close(reader)</code></pre><pre><code class="nohighlight hljs">┌ Warning: Reached end of reader; columns 7-25 are set to missing values
└ @ VCFTools /Users/biona001/.julia/dev/VCFTools/src/convert.jl:72</code></pre><p>As the warning suggests, the last window has less than 25 markers. The remaining columns in the matrix <code>g</code> are set to missing values.</p><h2 id="Sample-ID,-Chromosome,-SNP-position,-REF/ALT-alleles"><a class="docs-heading-anchor" href="#Sample-ID,-Chromosome,-SNP-position,-REF/ALT-alleles">Sample ID, Chromosome, SNP position, REF/ALT alleles</a><a id="Sample-ID,-Chromosome,-SNP-position,-REF/ALT-alleles-1"></a><a class="docs-heading-anchor-permalink" href="#Sample-ID,-Chromosome,-SNP-position,-REF/ALT-alleles" title="Permalink"></a></h2><p>To extract sample ID without looping over the entire VCF file, you can do</p><pre><code class="language-julia hljs">ids = sampleID(&quot;test.08Jun17.d8b.vcf.gz&quot;)</code></pre><pre><code class="nohighlight hljs">191-element Vector{String}:
 &quot;HG00096&quot;
 &quot;HG00097&quot;
 &quot;HG00099&quot;
 &quot;HG00100&quot;
 &quot;HG00101&quot;
 &quot;HG00102&quot;
 &quot;HG00103&quot;
 &quot;HG00104&quot;
 &quot;HG00106&quot;
 &quot;HG00108&quot;
 &quot;HG00109&quot;
 &quot;HG00110&quot;
 &quot;HG00111&quot;
 ⋮
 &quot;HG00383&quot;
 &quot;HG00384&quot;
 &quot;HG00403&quot;
 &quot;HG00404&quot;
 &quot;HG00406&quot;
 &quot;HG00407&quot;
 &quot;HG00418&quot;
 &quot;HG00419&quot;
 &quot;HG00421&quot;
 &quot;HG00422&quot;
 &quot;HG00427&quot;
 &quot;HG00428&quot;</code></pre><p>However, extracting each SNP&#39;s (record) chromosome, SNP position, or REF/ALT alleles, one must loop over the entire VCF file. This is achieved using the optional argument <code>save_snp_info = true</code> that can be supplied to functions <code>convert_gt</code>, <code>convert_ht</code>, and <code>convert_ds</code>.</p><pre><code class="language-julia hljs">@time X, X_sampleID, X_chr, X_pos, X_ids, X_ref, X_alt = convert_gt(Float64, 
    &quot;test.08Jun17.d8b.vcf.gz&quot;, save_snp_info=true)</code></pre><pre><code class="nohighlight hljs">  0.088368 seconds (536.26 k allocations: 43.550 MiB, 9.98% gc time)





(Union{Missing, Float64}[0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0; … ; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0], [&quot;HG00096&quot;, &quot;HG00097&quot;, &quot;HG00099&quot;, &quot;HG00100&quot;, &quot;HG00101&quot;, &quot;HG00102&quot;, &quot;HG00103&quot;, &quot;HG00104&quot;, &quot;HG00106&quot;, &quot;HG00108&quot;  …  &quot;HG00403&quot;, &quot;HG00404&quot;, &quot;HG00406&quot;, &quot;HG00407&quot;, &quot;HG00418&quot;, &quot;HG00419&quot;, &quot;HG00421&quot;, &quot;HG00422&quot;, &quot;HG00427&quot;, &quot;HG00428&quot;], [&quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;  …  &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;, &quot;22&quot;], [20000086, 20000146, 20000199, 20000291, 20000428, 20000683, 20000771, 20000793, 20000810, 20000814  …  20099406, 20099579, 20099654, 20099659, 20099660, 20099674, 20099716, 20099752, 20099891, 20099941], [[&quot;rs138720731&quot;], [&quot;rs73387790&quot;], [&quot;rs183293480&quot;], [&quot;rs185807825&quot;], [&quot;rs55902548&quot;], [&quot;rs142720028&quot;], [&quot;rs114690707&quot;], [&quot;rs189842693&quot;], [&quot;rs147349046&quot;], [&quot;rs183154520&quot;]  …  [&quot;rs41281429&quot;], [&quot;rs145947632&quot;], [&quot;rs9605066&quot;], [&quot;rs142467695&quot;], [&quot;rs74605905&quot;], [&quot;rs145967409&quot;], [&quot;rs139838034&quot;], [&quot;rs73389792&quot;], [&quot;rs1048659&quot;], [&quot;rs113958995&quot;]], [&quot;T&quot;, &quot;G&quot;, &quot;A&quot;, &quot;G&quot;, &quot;G&quot;, &quot;A&quot;, &quot;A&quot;, &quot;T&quot;, &quot;C&quot;, &quot;T&quot;  …  &quot;G&quot;, &quot;CCA&quot;, &quot;C&quot;, &quot;C&quot;, &quot;C&quot;, &quot;T&quot;, &quot;C&quot;, &quot;G&quot;, &quot;C&quot;, &quot;T&quot;], [[&quot;C&quot;], [&quot;A&quot;], [&quot;C&quot;], [&quot;T&quot;], [&quot;T&quot;], [&quot;G&quot;], [&quot;C&quot;], [&quot;C&quot;], [&quot;T&quot;], [&quot;C&quot;]  …  [&quot;C&quot;], [&quot;C&quot;], [&quot;T&quot;], [&quot;T&quot;], [&quot;T&quot;], [&quot;C&quot;], [&quot;G&quot;], [&quot;T&quot;], [&quot;G&quot;], [&quot;A&quot;]])</code></pre><pre><code class="language-julia hljs"></code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../filter/">« Filter</a><a class="docs-footer-nextpage" href="../write/">Write »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Sunday 1 January 2023 02:31">Sunday 1 January 2023</span>. Using Julia version 1.8.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
